patches/tModLoader/Terraria/ModLoader/GlobalNPC.cs

```diff
@@ -522,6 +523,16 @@ public virtual void DrawBehind(NPC npc, int index)
 		public virtual void SetupTravelShop(int[] shop, ref int nextSlot) {
 		}
 
+		/// <summary>
+		/// Allows you to set an NPC's biome preferences and nearby npc preferences for the NPC happiness system. Recommended to only be used with NPCs that have shops.
+		/// </summary>
+		/// <param name="npc"></param>
+		/// <param name="shopHelperInstance"></param>
+		/// <param name="primaryPlayerBiome"></param>
+		/// <param name="nearbyNPCsByType"></param>
+		public virtual void NPCHappiness(NPC npc, int primaryPlayerBiome, ref ShopHelper shopHelperInstance, ref bool[] nearbyNPCsByType) {
+		}
+
```

Comment:
Document parameters. What's with the last two parameters being by ref? The last parameter could be a readonlyspan.

Reply:
Resolved